### PR TITLE
Added docs on software validation procedures

### PR DESCRIPTION
**Pull request by henrikstranneheim**

### CODE REVIEW

**Comment by henrikstranneheim** [comment]
Here is a draft! Let's discuss

File: validation/README.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+### 3.5 Tests
+This dependent on how each tool or method is set-up. There should be adequate tests, data, environements to cover major, minor and patch updates. In the end, it should report if everything proceeded according to expectations or not. All tests need to pass so a simple check; “yes” or “no” will suffice. 
+
+## 4 Validation
+
+The validation needs to be documented in an automated way. Only for major updates or completely new methods do we intend to manually document using the test specification + implementation plan approach.
+
+For the rest of the updates it should be possible to track the information required through a combination of Git commits, TravisCI statuses, and an overall log that summaries results from each integration test run 
+
+### 4.1 Patch
```
Comment (hassanfa) [inline validation/README.md]: I think we should clarify how we should increment the version in patch numbers (i.e. major patch decimal increment, e.g. 0.1).

File: validation/README.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+It becomes difficult then to follow the same validation workflow used for developing lab methods if you want to update tools and methods several times per week. Therefore we need a way to do this in a safe and clearly documented way.
+
+## 2.1 Prerequisites
+In order for this to work we need to streamline how new/updated code is deployed. This means we need to automate as much of the process as possible. We also need to trust that the new code will behave as expected. It’s therefore required to run tests continuously whenever we change some code. This also need to be as automated as possible.
+
+## 2.2 Documentation
+All the code in our system should be version controlled where every change to the codebase is time stamped, commented on, and can be tracked to a specific Git commit. This is the most granular level of documentation but will suffice for most of the small updates.
+
+We also provide a structured versioning of the individual tools following [semantic versioning](https://semver.org/) conventions. Every tool is accompanied by a “change log” that summarises the most significant changes between two versions.
```
Comment (hassanfa) [inline validation/README.md]: Shouldn't we move this to "patch" section? see my comment below.

File: validation/README.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+   * MySQL (clinical-db)
+
+### 3.4 Data
+Test data should exist to start the integration tests from some key entry points:
+
+   * Demultiplexing from the BCL output of a flowcell
+   * Data analysis from linked FASTQ-files
+
+### 3.5 Tests
+This dependent on how each tool or method is set-up. There should be adequate tests, data, environements to cover major, minor and patch updates. In the end, it should report if everything proceeded according to expectations or not. All tests need to pass so a simple check; “yes” or “no” will suffice.
```
Comment (hassanfa) [inline validation/README.md]: My OCD is kicking in :-) ->This is* dependent.

File: validation/README.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+   * Demultiplexing from the BCL output of a flowcell
+   * Data analysis from linked FASTQ-files
+
+### 3.5 Tests
+This dependent on how each tool or method is set-up. There should be adequate tests, data, environements to cover major, minor and patch updates. In the end, it should report if everything proceeded according to expectations or not. All tests need to pass so a simple check; “yes” or “no” will suffice. 
+
+## 4 Validation
+
+The validation needs to be documented in an automated way. Only for major updates or completely new methods do we intend to manually document using the test specification + implementation plan approach.
```
Comment (hassanfa) [inline validation/README.md]: This is quite broad. Validation requires a validated output and expected output. Considering that tools and methods being updated all the time, the output of the validation needs to be quantified in way that it doesn't deviate too much from expected output. And we need instructions for the deviations ( I can't think of any good solution right now)

Comment (hassanfa) [reply validation/README.md]: Update: By deviations, I don't mean deviations in AMS style but rather in a more quantified way by defining certain clear statistical metrics. Otherwise validations will turn into just another unit test.

File: validation/README.md (hunk 1 above)
Comment (henrikstranneheim) [reply validation/README.md]: It is specified in the semantic versioning docs

File: validation/README.md (hunk 2 above)
Comment (henrikstranneheim) [reply validation/README.md]: This is a more general discussion across major, minor and patch

File: validation/README.md (hunk 3 above)
Comment (henrikstranneheim) [reply validation/README.md]: Should probably be "This is dependent on" 😄 . I will make correction

File: validation/README.md (hunk 4 above)
Comment (henrikstranneheim) [reply validation/README.md]: I agree, but it all relates to the type of update (major, minor, patch). If there is no expected change in the analysis output, unit/integration tests should be fine. We need to balance the docs and evaluation load with code maintenance and minor updates. Benchmark to a golden standard with a hapmap sample should cover most aspects of this.

File: validation/README.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+Then there’s a layer on top of the small tests that combine multiple tools and tests e.g. the entire workflow on a high level (integration tests). These tests don’t concern themselves so much with edge cases and are perhaps run in a production-like environment only once before new code is deployed to the system.
+
+### 3.1 Setup
+Unit tests are setup for each package/tool separately. They tend to use the py.test test-runner or a test anything protocol (TAP) and are automated using TravisCI. This makes it possible to track each result in the GitHub interface.
+
+It’s important that the environment used to run the integration tests behave as close to the production environment as possible. We achieve this by running the development code on the same hardware alongside the production code, but in isolated test environments.
+
+### 3.2 Installation
+All production code is installed in a conda environments following this naming [convention](http://localhost:4000/conda/conda_conventions.html):
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply validation/README.md]: Yeah, that seems to point in the wrong URL. I will fix it.

**Comment by henrikstranneheim** [comment]
I have now added a some instructions on how to set up a MIP testing/validation environment as well as common resources one have to use e.g. data set, configs etc. This is to facilitate a step wise instruction on how to test and validate, which can be as verbose as we want compared to the more A to B recipe in AM.

File: validation/README.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+We also provide a structured versioning of the individual tools following [semantic versioning](https://semver.org/) conventions. Every tool is accompanied by a “change log” that summarises the most significant changes between two versions.
+
+## 3 Testing
+There’s multiple levels of testing software. You need small, quick, and directed tests (unit tests) to make sure that you cover edge cases and stay clear of obvious coding errors (e.g. misspelling a variable). These tests can be run constantly and on any computer/server.
+
+Then there’s a layer on top of the small tests that combine multiple tools and tests e.g. the entire workflow on a high level (integration tests). These tests don’t concern themselves so much with edge cases and are perhaps run in a production-like environment only once before new code is deployed to the system.
+
+### 3.1 Setup
+Unit tests are setup for each package/tool separately. They tend to use the py.test test-runner or a test anything protocol (TAP) and are automated using TravisCI. This makes it possible to track each result in the GitHub interface.
```
Comment (moonso) [inline validation/README.md]: I think we should make it clear that we use pytest at clinical for python code

**Comment by henrikstranneheim** [comment]
Alright, I will merge this and we will keep updating this as we go along
